### PR TITLE
Fix detection of unknown base options in subproj

### PR DESCRIPTION
**Pull request by bruchar1**
cc4cfbcad92945a1629c80664e1eb755c68905dd added detection for unknown base options, but it was not working for subproject base options. This should fix it.

### CODE REVIEW

File: unittests/platformagnostictests.py [hunk 1]
```diff
@@ -282,7 +282,7 @@ def test_reconfigure_base_options(self):
         self.assertIn('\nMessage: b_ndebug: true\n', out)
         self.assertIn('\nMessage: c_std: c89\n', out)
 
-        out = self.init(testdir, extra_args=['--reconfigure', '-Db_ndebug=if-release', '-Dc_std=c99'])
+        out = self.init(testdir, extra_args=['--reconfigure', '-Db_ndebug=if-release', '-Dsub:b_ndebug=false', '-Dc_std=c99'])
         self.assertIn('\nMessage: b_ndebug: if-release\n', out)
         self.assertIn('\nMessage: c_std: c99\n', out)
```
Comment (dcbaker) [inline unittests/platformagnostictests.py:287]: Don't you need to add additional `self.assertIn` cases here?

File: mesonbuild/coredata.py [hunk 2]
```diff
@@ -1009,7 +1009,7 @@ def set_default_options(self, default_options: T.MutableMapping[OptionKey, str],
             # adding languages and setting backend.
             if k.type in {OptionType.COMPILER, OptionType.BACKEND}:
                 continue
-            if k.type == OptionType.BASE and k in base_options:
+            if k.type == OptionType.BASE and k.evolve(subproject='') in base_options:
```
Comment (dcbaker) [inline mesonbuild/coredata.py:1012]: There is a `k.as_root()`helper, but I don't feel strongly either way.

**Review by dcbaker — CHANGES_REQUESTED** [review]
The test doesn't seem to actually test that the option is properly set, which I'd like to see. Otherwise this looks like the right solution to me.

File: unittests/platformagnostictests.py (hunk 1 above)
Comment (bruchar1) [reply unittests/platformagnostictests.py:287]: You are right. I fixed the error causing the base option to be not recognised, but now I realise that the subproject option is still not set from the command line...

Comment (bruchar1) [reply unittests/platformagnostictests.py:287]: should be fixed now.

File: mesonbuild/coredata.py [hunk 3]
```diff
@@ -1016,15 +1016,24 @@ def set_default_options(self, default_options: T.MutableMapping[OptionKey, str],
...
+    def add_compiler_options(self, options: MutableKeyedOptionDictType, lang: str, for_machine: MachineChoice,
+                             env: Environment, subproject: str) -> None:
         for k, o in options.items():
             value = env.options.get(k)
             if value is not None:
                 o.set_value(value)
-                self.options[k] = o  # override compiler option on reconfigure
+                if not subproject:
+                    self.options[k] = o  # override compiler option on reconfigure
             self.options.setdefault(k, o)
```
Comment (dcbaker) [inline mesonbuild/coredata.py:1027]: This set's the subproject='' key to the subproject value, that that doesn't seem right.

**Review by dcbaker — APPROVED** [review]
Okay, it took me a few minutes to parse everything that was giong on here, but I think this is right.

**Pull request merged**